discovery_engine_core/core/src/stack/filters.rs

```diff
@@ -39,52 +36,57 @@ impl ArticleFilter for DuplicateFilter {
     fn apply(
         history: &[HistoricDocument],
         stack: &[Document],
-        articles: Vec<Article>,
+        mut articles: Vec<Article>,
     ) -> Result<Vec<Article>, GenericError> {
-        let urls = history
+        let urls = stack
             .iter()
-            .map(|doc| doc.url.as_str())
-            .chain(stack.iter().map(|doc| doc.resource.url.as_str()))
-            .collect::<HashSet<_>>();
-
-        let titles = history
+            .map(|doc| (doc.resource.url.as_str(), doc.resource.rank))
+            .chain(history.iter().map(|doc| (doc.url.as_str(), 0)))
+            .fold(HashMap::new(), |mut urls, (url, rank)| {
+                let best_rank = urls.entry(url).or_insert(rank);
+                if rank < *best_rank {
+                    *best_rank = rank;
+                };
+                urls
+            });
+
+        let titles = stack
             .iter()
-            .map(|doc| &doc.title)
-            .chain(stack.iter().map(|doc| &doc.resource.title))
-            .collect::<HashSet<_>>();
-
-        let mut articles: BTreeSet<_> = articles.into_iter().map(UniqueArticle).collect();
-
-        articles.retain(|article| {
-            !(urls.contains(&article.0.link.as_str()) || titles.contains(&&article.0.title))
+            .map(|doc| (&doc.resource.title, doc.resource.rank))
+            .chain(history.iter().map(|doc| (&doc.title, 0)))
+            .fold(HashMap::new(), |mut titles, (title, rank)| {
+                let best_rank = titles.entry(title).or_insert(rank);
+                if rank < *best_rank {
+                    *best_rank = rank;
+                };
+                titles
+            });
+
+        // discard dups in the title keeping only the best ranked
```

Comment:
🟢  we could reorder the code blocks such that first the title map & dedup is done, and afterwards the url map & dedup is done. this would place the map next to the dedup for better readability, also the compiler could drop the maps sooner. (the retaining by rank would come as a third block like it currently is.)

Reply:
thanks. actually the `titles` and `urls` hash maps are used by the "retaining" block (rather than the dedup block) so what I'll do is move them down so they're closer together.